[WIP] Initial work

AuHau opened the pull request:
A first draft version of the migration tool, that can be meaningfully feedbacked.

## Scope-in

Scope and status of this PR (and package):
- [x] Define the migration's API
- [x] Create high-level API to migrate forward&backward migrations
- [x] Create CLI to run the migrations
- [x] Write documentation to README
    - [ ] Document tests, how to write them, run them etc
- [ ] Write test coverage

## Scope-out 

Thinks that are not in scope of this PR and will be scope of separate PR to js-ipfs that will integrate this package:
 - Disabling/enabling automatic migrations 
 - Propagating information about migrations to JS-IPFS readme
 - Integrate to js-ipfs with posibility to automatically run the migrations

## Problems

 1. If the automatic migrations will be disabled in browser environment, how does the user can manually migrate its repo, when it does not have access to CLI?

AuHau commented:
So the PR is ready for review. I would also ask for the guidance of what next steps should be (ofc after reviews)? I assume something like:

* Move repo under IPFS org
* License file move to Protocol Labs?
* Integrate CI
* Release

As a community member, I can't do much of these, but I would love to work with you, so you could just hire me and then I could do them! 😉 

### Blocked by release of async/await versions of 
 * [x] interface-datastore
 * [x] js-datastore-fs
 * [x] js-datastore-level

dirkmc commented on src/errors.js:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Exception raised when trying to revert migration that is not possible
+ * to revert.
+ */
```
For these errors can we use an [error code](https://github.com/IndigoUnited/js-err-code#readme) instead of a name, and add tests that cause these errors to be thrown and check the error code. Here's an [example of a test that checks an error code](https://github.com/ipfs/js-datastore-s3/blob/master/test/index.spec.js#L163)

dirkmc commented on test/index.spec.js:
```diff
@@ -0,0 +1,380 @@
+      const migrationsMock = createMigrations()
+
+      return expect(migrator.revert(undefined, undefined, undefined, undefined, undefined, undefined, migrationsMock))
+        .to.eventually.be.rejectedWith('Path argument is required!')
```
Let's add error codes for these too

AuHau commented:
Hey there!
Sweet, thanks for looking into it!

To your comments.

Ad 1. Of course, I can trim down the CLI to follow the same signature like `fs-repo-migration`, but to be honest, its CLI is very basic and I would say could be improved. I assume this request originates in the interopt endeavor with go-ipfs? It is a question of how this component can be interoperable because, in the beginning, it will definitely won't be as there are still differences between js/go and that will require different migrations. Which sort-of imply that these package will be always only for `js-ipfs` repository and not for `go-ipfs` one. Soooooo do you want me still to trim it down? 😇 

Ad 2. Alright, will do.

Ad 3. Yeah, I was thinking about it, but for some reason didn't go this way (most probably because in the beginning there were not so many optional parameters). Is it alright to still keep required parameters (for `migrate()` => `path` and for `revert()` => `path` and `version`) as parameters and only the optional one move to an object? Regarding the path parameter. I believe that it is not in the scope of the migration function figuring out where the data are placed. The higher component should be aware of this (eq. `js-ipfs` or the CLI interface) and hence should provide this information to these functions. 

Ad 4. Not sure exactly what you mean. Do you want me to create folder `example` and put an example migration there? (But it won't be hooked up into the "live" version)?

dirkmc commented:
1. I think it's fine to have more features than the go version of the repo migration tool. I'm just saying let's try to make sure the command and parameter names match where the functionality is the same, eg `ver` -> `version`

2. 👍 

3. That sounds good to me, I understand now why path is a mandatory parameter.

4. I'm suggesting that you add some simple migrations as part of the tests, eg changing a config variable name. This would allow us to write a kind of integration test that runs realistic migrations and check that it worked. For example
- a migration v1-v2 that changes a config variable name `myvar` to `mynewvar`
- a migration v2-v3 that changes the config variable name `mynewvar` to `mynewvar2`
You could then create tests that would run the migration from v1 to v2, v2 to v3, v1 to v3 and the reverse, and check the actual config file itself to ensure that the migration ran correctly.
I'm not sure if I'm being clear, does that make sense?

AuHau commented:
1. Okey! Good! Well, the thing is that I use subcommands, but `go` does not. Their signature is `fs-repo-migrations [-to <int>] [-v] [-y]`, but mine signature is `ipfs-repo-migrations migrate|reverse|status|add ...`. Sooo should I drop the subcommands and go something like: `ipfs-repo-migrations [--to <int>] [--status] [--add]`? My main question here is, does reversion (eq. backward migration) needs to be separated from forward migration? Eq. does the user needs to be wary that he is reverting? After some thinking, I guess not. He only cares about having the repo at a specific version but does not care if the tool has to revert or migrate.

2. 👍 
3. 👍 
4. 👍, make sense will do.

dirkmc commented:
fs-repo-migrations has the following helptext:
```
$ fs-repo-migrations --help
Usage of fs-repo-migrations:
  -revert-ok
    	allow running migrations backward
  -to int
    	specify version to upgrade to (default 7)
  -v	print highest repo version and exit
  -y	answer yes to all prompts
```

I wasn't able to install js-ipfs-repo-migrations on this (dev) branch because of some badly linked dependencies (in underlying deps), could you please show me what the helptext for the command structure you're proposing would look like? (roughly)

Also please go ahead and update the dependencies, they should all be released now I believe.

AuHau commented:
Dependencies updated.

Current help text (top-level), don't mind the `cli.js`:
```
cli.js <command>

Commands:
  cli.js migrate       Migrate IPFS repo to latest or specific version
  cli.js revert <ver>  Revert IPFS repo to specific version
  cli.js status        Display status of IPFS repo
  cli.js add           Bootstrap new migration

Options:
  --help       Show help                                               [boolean]
  --version    Show version number                                     [boolean]
  --repo-path  Path to the IPFS repo                                    [string]
```

Then for example `migrate --help`:
```
cli.js migrate

Migrate IPFS repo to latest or specific version

Options:
  --help       Show help                                               [boolean]
  --version    Show version number                                     [boolean]
  --repo-path  Path to the IPFS repo                                    [string]
  --ver        Specify to which version should be repo migrated to      [number]
  --dry        Only displays what migrations will be migrated          [boolean]
```

AuHau commented:
If we would want to drop the sub-commands approach, then I guess we could do something like:

```
cli.js

Migrate/reverse IPFS repo to specific version

Options:
  --help       Show help                                               [boolean]
  --version    Show version number                                     [boolean]
  --repo-path  Path to the IPFS repo                                    [string]
  --to        Specify to which version should be repo migrated/reversed to      [number]
  --dry        Only displays what migrations will be reverted          [boolean]
  --status     Shows the current version of the repository and what migrations are available          [boolean]
  --add		   Bootstraps a new migration							   [boolean]
  --empty	   When --add, specifices to create empty migration		   [boolean]
```

dirkmc commented:
I see, I think I like the subcommands approach better actually, it's clearer. I think we should match parameter names where possible though.

It looks like there's a parsing error with package.json

AuHau commented:
Hmmmm, not sure where there is a problem? I pushed updated `package-lock.json`, maybe give it a try again? If there is still a problem, can you post more details?

Alright! Yeah, I like the subcommands approach more too. What are your thoughts on merging `migrate` and `revert` to one subcommand and detect what is needed based on the current version of the repo? If so, should there be similar parameter like in `go`: `--revert-ok`? I would maybe argue that it is not necessary?

To be honest, I am not sure what parameters should we match. The only thing that I see we could match is the go's `-to`, but tbh I would like then to match it to `--to` as that follows the general CLI guidelines that single `-` is used for single character shortcuts (eq. `-t`) and for word parameters is `--` used.

Edit: Heh, I think `-to` is not even possible with `yargs`...

dirkmc commented:
I think you resolved the conflict with your last push 👍 

I think it's fine to use idiomatic parameter parsing, eg use `--to` instead of `-to`. Otherwise I'd suggest that we mirror the parameters in js-repo-migrations but under the migrate subcommand, eg
`node cli.js migrate --to 7 --revert-ok`

I think it's probably safer to have a `--revert-ok` option

AuHau commented:
Alright, sweet! Yes, agree and all clear. I will implement all the remarks today 👍

AuHau commented:
Alright! Sweet! All clear, I will worked it all in today.

EDIT: Hehe, email replies take long time to propagate :-D 

On Thu, May 30, 2019, 11:33 dirkmc <notifications@github.com> wrote:

> I think you resolved the conflict with your last push 👍
>
> I think it's fine to use idiomatic parameter parsing, eg use --to instead
> of -to. Otherwise I'd suggest that we mirror the parameters in
> js-repo-migrations but under the migrate subcommand, eg
> node cli.js migrate --to 7 --revert-ok
>
> I think it's probably safer to have a --revert-ok option
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/AuHau/js-ipfs-repo-migrations/pull/1?email_source=notifications&email_token=ABOKPOV37FO4VXS4II6ZOS3PYAMXRA5CNFSM4HDXTDDKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWTENCA#issuecomment-497436296>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABOKPOUVC7Z7J65IETZN4ZDPYAMXRANCNFSM4HDXTDDA>
> .
>

AuHau commented:
Sooo I have worked on everything, except the "fixture migration test", which is WIP, but I have to focus on something else atm. So review it and let me know if there are more things to change.

I have not updated the `js-ipfs` integration PR, I will leave that as last thing after the work here is finished.

dirkmc commented:
Great, thank you Adam 👍 

I will take a look over the weekend

dirkmc commented on src/errors.js:
```diff
@@ -4,35 +4,75 @@
   constructor (message) {
     super(message)
-    this.name = 'NonReversibleMigration'
+    this.name = 'NonReversibleMigrationError'
```
I don't think it's necessary for these Errors to have name field.

AuHau replied:
I think it is actually good idea 😉 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/name

dirkmc replied:
Fair enough 👍

dirkmc commented on src/errors.js:
```diff
@@ -4,35 +4,75 @@
 }
-exports.NonReversibleMigration = NonReversibleMigration
+
+NonReversibleMigrationError.code = 'ERR_NON_REVERSIBLE_MIGRATION'
```
We probably don't need to define the code as a class field as it is already declared in the constructor

AuHau replied:
If you want to assert that some error is being thrown with an error code of the error, then yes you need it because if it is declared only in constructor then you don't have any way how to retrieve without instantiating it and hence `if (err.code == errors.NonReversibleMigrationError.code) ...` not possible.

AuHau replied:
Btw. this is pretty much same approach that is proposed by @hugomrdias here: https://github.com/ipfs/js-ipfs/pull/1746/files#diff-00052fb38502dc1ed2fbc79ca57f9e4e

dirkmc replied:
Ok makes sense 👍

dirkmc commented on test/index.spec.js:
```diff
@@ -80,22 +80,22 @@ describe('index.js', () => {
     expect(
       () => migrator.getLatestMigrationVersion([])
-    ).to.throw('Migrations must be non-empty array!')
+    ).to.throw(errors.InvalidValueError, 'Migrations must be non-empty array!').with.property('code', errors.InvalidValueError.code)
```
Let's just test the code field (not the text of the error message itself) so that we can easily change the text later without having to change the tests

AuHau replied:
Yeah that is good idea, will fix this.

AuHau commented:
I will add tomorrow the fixture integration test and remove the error message assertions. Let me know if there is something else, otherwise, I will also update the `js-ipfs` integration PR...

dirkmc commented:
I would say let's hold off on updating the js-ipfs integration PR until this one is complete, as this review will probably require a few iterations. Writing tests may cause you to refactor some things, so I will continue reviewing once you have completed the tests.

AuHau commented:
So comments worked in.

I am not quite sure if the test migration is something you were looking for as it is not much simpler then the #2 one 😅 Let me know...

Also, the test fails in the Browser, because of the loading fixture. Have you used `loadFixture()` before? Is it correctly used? Does the path to the fixture look OK? [Code](https://github.com/AuHau/js-ipfs-repo-migrations/pull/1/files#diff-77686aae68ae4b50e2efc518d77fd6d3R26)

Also, I was thinking about adding `sharness` tests for the CLI. Do you think it is worth?

dirkmc commented:
Thanks @AuHau, these tests look great. Can we add tests for migrating more than one version forward and backwards as well?

Here is an example of how to use fixtures: https://github.com/ipfs/interface-js-ipfs-core/blob/e1611757b868ca0e308c531b42467f168be351a3/src/pin/utils.js
I'm not 100% sure but I believe they may need to be under the directory `test/fixtures`

I believe that the go team is moving away from using sharness tests, I will try to find out why.

It is possible to use package.json files instead of storing all the code for node modules that have changed?

AuHau commented:
Sweet! 

Well of course I can! But I think the logic of migration multiple tests is quiet covered with the `index.spec.js` tests, like [here](https://github.com/AuHau/js-ipfs-repo-migrations/pull/1/files#diff-faa503f61f8addeb603095cc2f40c82eR150). So is there something specific you want me to cover with these tests? Or is the `index.spec.js` coverage enough?

Hmm, that sounds strange that it would have to be in a specific folder, but I am out of ideas so I will try it.

Oh, I see, good to know!

Hmm, I am not sure if I get your point. What are you referring to with "node modules that have changed"? Are you referring to the point that `node_modules` are committed to the repo? If so, that is a principle copied from `fs-repo-migrations` and makes sense to me as it tries to guarantee that the migration will always work as there is some chance that in future a package can disappear from the registry, which would break the migrations.

dirkmc commented:
Ok I see that there is good coverage for traversing multiple migration versions, maybe it makes sense to keep the integration tests simple 👍 

It looks like the fixtures do need to be in that specific folder: https://github.com/ipfs/aegir#fixtures

I was referring to committing node modules to the repo. The only reason I can think of that a package would disappear from the npm registry is because it was taken down for security reasons, in which case we probably don't want to retrieve it. Is there another reason you're aware of?

AuHau commented:
👍 Sweet!

Alright, will take look into it.

Hmmmmm, tbh I don't know `npm` registry that well. Is it possible for the author to delete the package? It does not seem so, right? Once it is published it can't be taken down unless whole `npm` is taken down, right? For example, with PyPi you can remove the package from the registry and such property I feel makes it "unstable" enough for this sort of use case. That said I am not 100% sure how dependency management in Go works, but I think it is very tightly coupled with GitHub repositories and there is no "registry", so the original `fs-repo-migrations`'s rule about copying dependencies directly into the repo is maybe in a more different context that we are in. What do you think?

dirkmc commented:
That sounds correct to me. I would suggest we go with the solution that relies on npm being available, as it is central to the Javascript ecosystem so I think that's a reasonable assumption.

I asked about sharness and it seems like the go project is still using them, so I think I was mistaken. I like the idea of having some simple sharness tests.

AuHau commented:
Alright, agree. But then, how we will handle multiple `package.json`? Do you have experience with it? After some basic research, it does not seem like a simple thing to do as it will require some extra tooling around merging all the dependencies into one root `package.json`. Maybe easier would be to have just one? I don't think it really makes sense to have multiple ones...

Alright, I will have a look into some basic sharness tests.

dirkmc commented:
I'm not sure I understand the question.. my understanding is that a particular package.json will be associated with a particular js-repo version. So for example to migrate to repo version 7, migration 7 would need to include a package.json, and the migrator would load the package.json for migration 7 and run `npm install`

AuHau commented:
Hmmm, it seems like a quite complicated thing and also if I have correct understanding what you mean, not really possible.

First of all, I used multiple `package.json` only because it easily created the `node_modules` folder with the migration's dependencies that I could commit (I did not want to commit the package's own dependencies - eq. root `node_modules`, but only the migration's dependencies). There is no other reason for it.

Regarding "migrator would load the package.json for migration 7 and run npm install
", that would imply that every time you run migrations there would be `npm install` invoked? That is not really possible in browser environments...

Since this package will be normally released on npm (I assume), I think that all the migration's dependencies have to be either committed (but we said no to that) or stated as the package itself dependencies.

dirkmc commented:
Oh right that's true, I didn't think about browsers, let me think about this a little more

AuHau commented:
Sure! I am soon leaving for a flight to Europe so I will be offline, but let me know your thoughts! I will get back to it hopefully tomorrow.

I guess the main concern here is the size of all dependencies and eventually the size of the bundle right? On one hand, I feel like most of the dependencies will be shared with `js-ipfs`, but the problem is that the versions of these packages will eventually differ as the migration's dependencies won't be updated and that will increase the size... Tbh not really sure what to do about it.

dirkmc commented:
That's part of the concern. It just feels a little strange to me to be committing node modules. When you run `npm install` it doesn't just copy code, it can also [run scripts](https://docs.npmjs.com/misc/scripts) preinstall or postinstall, and compile modules with bindings, which will differ on different operating systems.

On the other hand, as you pointed out, we need to be concerned with how migrations should work in browsers.

So maybe the correct solution here is to
- run `npm install` in a nodejs environment
- include a browser-ready distribution of the package for browsers

What do you think?

dirkmc commented:
PS I hope you had a good flight 🚀  :)

AuHau commented:
Hmmmm, I did not really think of the other aspects of `npm install`, but tbh isn't `js-ipfs` generally restricted to dependencies, that do not compile modules? As it has to run in browsers? 

Also, I feel this is a bit counterproductive approach because I would say that bundle size is mainly issue for browser environment, not really for `node` one right? (eq. `node` install the js-ipfs only once - or with updates, while in browser it is part of every page load) And since we will have browser bundle with all the dependencies anyway, it won't really improve anything on the `node` (with the approach of `npm install` on migration).

Also just to clarify, I am not proposing to commit `node_modules` anymore, just to merge all `package.json` (eq. the root one and all the migration's one) into the root one.

PS.: Flight was completely full, so no space to move 🙈

dirkmc commented:
I want to understand your proposal for merging all the package.json files. Could you write it out for me step by step, with an example?

AuHau commented:
Hey @dirkmc,
I am sorry that I went silent, but I am leaving for two weeks thru-hike to Corsica tomorrow and I have been preparing for it the last couple of days. I am afraid I won't be able to come back to this before the IPFS Camp, but we can have a chat about it there (I assume you are going there?) and maybe even finish it! 😉

dirkmc commented:
Sounds amazing! Thank you for all your hard work so far, I think we're very close to the finish line for this new project.

Unfortunately I can't make it to IPFS camp this year, but there will be plenty of experts there, it is a high priority for us to get done by the end of the quarter, so please do ask for feedback from the JS IPFS team, and I will give them a heads up as well.

Enjoy the trip!

AuHau commented:
Oh that is pity, that we won't meet in person, but I understand. Alright, I will talk with people there to push it into the finish. 

Thanks for all the feedback so far!

AuHau commented:
Hey there @dirkmc

So I have discussed this project with @alanshaw. He gave me some more feedback and also we reached the solution for the migration's dependencies. As it is important to keep the `js-ipfs` bundle small we agreed that the dependencies will have to be kept  in sync with `js-ipfs` (if they overlap). You can find some discussion I wrote about this [here](https://github.com/AuHau/js-ipfs-repo-migrations/pull/1/commits/f5f3ddff65c94653a845535a14d33e60604127d3#diff-04c6e90faac2675aa89e2176d2eec7d8R166).

Alan's feedback is already all implemented (it was few small things). I have prepared the `sharness` tests (but with no real tests yet), so I will finish those and still have to finish the integration tests for browsers (eq. loading fixtures). But otherwise I think it is pretty close to finish line.

Do you have some more comments from your side? 

Otherwise I have also reworked the original `js-ipfs` integration PR with moving the actual integration to `js-ipfs-repo` as I think it makes more sense: https://github.com/ipfs/js-ipfs-repo/pull/202

dirkmc commented:
That's great, thank you very much!

I don't have any more feedback for now, I will review once you have added the sharness tests.

AuHau commented:
Sorry @dirkmc that it took me so long, but now I believe everything is finished. Sharness tests and also working integration tests.

With Alan we were discussing if it would make sense to introduce versioning scheme of this package like this: `<migrationVersion>.<patch>`, where the `migrationVersion` is number of highest migration and `patch` is for fixes. It would be easy to see what version is then used. But on the other hand it does not follow the standard SemVer... What are your thoughts on it?

dirkmc commented:
That's great news, thanks! I will take a look on Monday.

hmm that's a good question. Maybe we could use semver but the migration version is the minor and patch is patch?

eg `0.<migrationVersion>.<patch>`

AuHau commented:
Well, that is definitely an option, but I am not sure if it makes sense? I guess this connects to the semantic of versioning. So your proposed scheme would signal that there are never breaking changes which the other scheme I guess could invoke. But at the same time to prefix something with zero, just to make it look the same does not feel completely good approach too. But I guess considering that all (? I guess?) IPFS's packages follow SemVer it would be weird not to. So I am down for your proposition.

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+> Migration framework for versioning of JS IPFS Repo
+
+This package takes inspiration from similar tool used by Go-IPFS: [fs-repo-migrations](https://github.com/ipfs/fs-repo-migrations/)
```
```suggestion
This package is inspired by the [go-ipfs repo migration tool](https://github.com/ipfs/fs-repo-migrations/)
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+## Background
+
+As JS-IPFS evolves and new technologies, algorithms and data structures are being incorporated it is necessary to 
```
```
As js-ipfs evolves and new technologies, algorithms and data structures are incorporated it is necessary to 
enable users to transition between versions. Different versions of js-ipfs may expect a different IPFS repo structure or content (see: [IPFS repo spec](https://github.com/ipfs/specs/tree/master/repo), [JS implementation](https://github.com/ipfs/js-ipfs-repo) ).
So the IPFS repo is versioned, and this package provides a framework to create migrations to transition
from one version of IPFS repo to the next/previous version.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+For that reason IPFS repo is versioned and this package provides framework to create migrations which transits
+one version of IPFS repo into next/previous one.
+
+This framework provides:
```
```suggestion
This framework:
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+This framework provides:
+ * Handles locking/unlocking of repository
+ * Defines migrations API
+ * Executes and reports migrations in both direction: forward and backward
```
```suggestion
 * Executes and reports migrations in both directions: forward and backward
```

dirkmc commented on README.md:
````diff
@@ -1 +1,284 @@
+}
+```
+
+On how to migrate your repository using CLI, you can find the recommended steps in [how to run migrations](./run.md) tutorial. 
````
```suggestion
To migrate your repository using the CLI, see the [how to run migrations](./run.md) tutorial. 
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+On how to migrate your repository using CLI, you can find the recommended steps in [how to run migrations](./run.md) tutorial. 
+
+**For tools that build on top of `js-ipfs` and run mainly in the browser environment, be aware of disabling automatic
```
**For tools that build on top of `js-ipfs` and run mainly in the browser environment, be aware that disabling automatic
migrations leaves the user with no way to run the migrations because there is no CLI in the browser. In such
a case, you should provide a way to trigger migrations manually.**

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+#### Architecture of migrations
+
+All migrations are placed in `/migrations` folder. Each folder there represents one migration that follows migration
```
```suggestion
All migrations are placed in the `/migrations` folder. Each folder there represents one migration that follows the migration
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+All migrations are placed in `/migrations` folder. Each folder there represents one migration that follows migration
+API.
+
+All migrations are collected in `/migrations/index.js`, which should not be edited manually is it is regenerated on
```
All migrations are collected in `/migrations/index.js`, which should not be edited manually. It is regenerated on
every run of `jsipfs-migrations add` (manual changes should follow the same style of modifications). 
**The order of migrations is important and migrations must be sorted in ascending order**.

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+every run of `jsipfs-migrations add` (or the manual changes should follow same style of modifications). 
+**The order of migrations is important and the migrations have to be sorted in the growing order**.
+
+Each migration has to follow this API. It has to export object in its `index.js` that has following properties:
```
```suggestion
Each migration must follow this API. It must export an object in its `index.js` that has following properties:
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+Each migration has to follow this API. It has to export object in its `index.js` that has following properties:
+
+ * `version` (int) - Number that represents the version into which will be the repo migrated to (eq. `migration-8` will move the repo into version 8).
```
```suggestion
 * `version` (int) - Number that represents the version which the repo will migrate to (eg. `migration-8` will move the repo to version 8).
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+ * `version` (int) - Number that represents the version into which will be the repo migrated to (eq. `migration-8` will move the repo into version 8).
+ * `description` (string) - Brief description of what the migrations does.
+ * `migrate` (function) - Function that on execution will perform the migration, see signature of this function bellow.
```
```suggestion
 * `migrate` (function) - Function that performs the migration (see signature of this function below)
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+ * `version` (int) - Number that represents the version into which will be the repo migrated to (eq. `migration-8` will move the repo into version 8).
+ * `description` (string) - Brief description of what the migrations does.
+ * `migrate` (function) - Function that on execution will perform the migration, see signature of this function bellow.
+ * `revert` (function) - If defined then this function will be used to revert the migration to previous version. Otherwise it is assumed that it is not possible to revert this migration.
```
```suggestion
 * `revert` (function) - If defined then this function will revert the migration to the previous version. Otherwise it is assumed that it is not possible to revert this migration.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+Arguments:
+ * `repoPath` (string) - absolute path to the root of the repo
+ * `options` (object, optional) - object containing `IPFSRepo` options, that should be used to construct datastore instance.
```
```suggestion
 * `options` (object, optional) - object containing `IPFSRepo` options, that should be used to construct a datastore instance.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+Arguments:
+ * `repoPath` (string) - absolute path to the root of the repo
+ * `options` (object, optional) - object containing `IPFSRepo` options, that should be used to construct datastore instance.
+ * `isBrowser` (bool) - indicates if the migration is run in browser environment in oppose to NodeJS
```
```suggestion
 * `isBrowser` (bool) - indicates if the migration is run in a browser environment (as opposed to NodeJS)
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+Arguments:
+ * `repoPath` (string) - path to the root of the repo
+ * `options` (object, optional) - object containing `IPFSRepo` options, that should be used to construct datastore instance.
```
```suggestion
 * `options` (object, optional) - object containing `IPFSRepo` options, that should be used to construct the datastore instance.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+Arguments:
+ * `repoPath` (string) - path to the root of the repo
+ * `options` (object, optional) - object containing `IPFSRepo` options, that should be used to construct datastore instance.
+ * `isBrowser` (bool) - indicates if the migration is run in browser environment in oppose to NodeJS
```
```suggestion
 * `isBrowser` (bool) - indicates if the migration is run in a browser environment (as opposed to NodeJS)
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+#### Browser vs. NodeJS environments
+
+Migration might need to distinguish in what environment it runs (browser vs. NodeJS), for this reason there is the argument
```
```suggestion
The migration might need to distinguish in which environment it runs (browser vs. NodeJS). For this reason there is an argument
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+Migration might need to distinguish in what environment it runs (browser vs. NodeJS), for this reason there is the argument
+`isBrowser` passed to migrations functions. But with simple migrations it should not be necessary to distinguish between
+these environments as datastore implementation will handle the main differences. 
```
```suggestion
these environments as the datastore implementation will handle the main differences. 
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+these environments as datastore implementation will handle the main differences. 
+
+There are currently two main datastore implementations:
+ 1. [`datastore-fs`](https://github.com/ipfs/js-datastore-fs) that is backed by file system and is used mainly in NodeJS environment
```
```suggestion
 1. [`datastore-fs`](https://github.com/ipfs/js-datastore-fs) that is backed by file system and is used mainly in the NodeJS environment
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+There are currently two main datastore implementations:
+ 1. [`datastore-fs`](https://github.com/ipfs/js-datastore-fs) that is backed by file system and is used mainly in NodeJS environment
+ 2. [`datastore-level`](https://github.com/ipfs/js-datastore-level) that is backed by LevelDB and is used mainly in browser environment
```
```suggestion
 2. [`datastore-level`](https://github.com/ipfs/js-datastore-level) that is backed by LevelDB and is used mainly in the browser environment
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+ 1. [`datastore-fs`](https://github.com/ipfs/js-datastore-fs) that is backed by file system and is used mainly in NodeJS environment
+ 2. [`datastore-level`](https://github.com/ipfs/js-datastore-level) that is backed by LevelDB and is used mainly in browser environment
+ 
+ Both implementations share same API and hence are interchangeable. 
```
```suggestion
 Both implementations share the same API and hence are interchangeable. 
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+ 
+ Both implementations share same API and hence are interchangeable. 
+ 
+ When the migration is run in browser environment the `datastore-fs` is automatically replaced with `datastore-level` even
```
```
When the migration is run in a browser environment, `datastore-fs` is automatically replaced with `datastore-level` even when it is directly imported (`require('datastore-fs')` will return `datastore-level` in a browser). So with simple migrations you shouldn't worry about the difference between `datastore-fs` and `datastore-level` and by default use the `datastore-fs` package (as the replace mechanism does not work vice versa).
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+#### Guidelines
+
+The recommended way to write a new migration is to first bootstrap an dummy migration using the CLI:
```
```suggestion
The recommended way to write a new migration is to first bootstrap a dummy migration using the CLI:
```

dirkmc commented on README.md:
````diff
@@ -1 +1,284 @@
+> jsipfs-migrations add
+```
+
+Afterwards new folder is created with bootstrapped migration. You can then simply fill in the required fields and 
````
```suggestion
A new folder is created with the bootstrapped migration. You can then simply fill in the required fields and 
```

dirkmc commented on README.md:
````diff
@@ -1 +1,284 @@
+```
+
+Afterwards new folder is created with bootstrapped migration. You can then simply fill in the required fields and 
+write the rest of migration! 
````
```suggestion
write the rest of the migration! 
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+Afterwards new folder is created with bootstrapped migration. You can then simply fill in the required fields and 
+write the rest of migration! 
+
+The `node_modules` of the migration should be committed to the repo to ensure that the dependencies are resolved even in
```
Is this still true?

AuHau replied:
Nope, you are right, I missed this statement when I was rewriting this aspect. Thanks!

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+#### Migration's dependencies
+
+Size of `js-ipfs`'s bundle is crucial for distribution in browser environment, hence dependency management of all related
```
```suggestion
The size of the `js-ipfs` bundle is crucial for distribution in a browser environment, so dependency management of all related
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+Size of `js-ipfs`'s bundle is crucial for distribution in browser environment, hence dependency management of all related
+packages is important.
+
+If migration need to depend on some package, this dependency should be declared in the root's `package.json`. Author
```
```suggestion
If a migration needs to depend on some package, this dependency should be declared in the root's `package.json`. The author
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+packages is important.
+
+If migration need to depend on some package, this dependency should be declared in the root's `package.json`. Author
+of migrations should be thoughtful about adding dependencies that would significantly increase the size of the final bundle.
```
```suggestion
of the migration should be thoughtful about adding dependencies that would significantly increase the size of the final bundle.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+of migrations should be thoughtful about adding dependencies that would significantly increase the size of the final bundle.
+
+Most of the migration's dependencies will most likely overlap with `js-ipfs`'s dependencies and hence should not introduce
+any significant overhead, but it requires to keep the versions of these dependencies in sync with `js-ipfs`. For this 
```
```suggestion
any significant overhead, but it is necessary to keep the versions of these dependencies in sync with `js-ipfs`. For this 
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+Most of the migration's dependencies will most likely overlap with `js-ipfs`'s dependencies and hence should not introduce
+any significant overhead, but it requires to keep the versions of these dependencies in sync with `js-ipfs`. For this 
+reason migrations should be well tested to ensure correct behaviour over its dependencies updates. It might happen
```
```suggestion
reason migrations should be well tested to ensure correct behaviour over dependency updates.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+Most of the migration's dependencies will most likely overlap with `js-ipfs`'s dependencies and hence should not introduce
+any significant overhead, but it requires to keep the versions of these dependencies in sync with `js-ipfs`. For this 
+reason migrations should be well tested to ensure correct behaviour over its dependencies updates. It might happen
+that update of some dependency could introduce breaking change, in such a case the next steps should be discussed with broader 
```
```suggestion
An update of some dependency could introduce breaking change. In such a case the next steps should be discussed with a broader 
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+#### Integration with js-ipfs
+
+When new migration is created, the repo version in [`js-ipfs-repo`](https://github.com/ipfs/js-ipfs-repo) should be updated with the new version,
```
```suggestion
When a new migration is created, the repo version in [`js-ipfs-repo`](https://github.com/ipfs/js-ipfs-repo) should be updated with the new version,
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+#### Integration with js-ipfs
+
+When new migration is created, the repo version in [`js-ipfs-repo`](https://github.com/ipfs/js-ipfs-repo) should be updated with the new version,
+together with updated version of this package. And afterwards propagate the updated version to `js-ipfs`.
```
```suggestion
together with updated version of this package. Then the updated version should be propagated to `js-ipfs`.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+#### Tests
+
+If migration affects working of any of the following functionality, it has to provide tests for the following functions
```
```suggestion
If a migration affects any of the following functionality, it must provide tests for the following functions
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+* `/src/repo/lock.js`:`lock()` - locking repository that uses file system
+* `/src/repo/lock-memory.js`:`lock()` - locking repository that uses memory
+
+Every mgration has to have a test coverage. Tests for migration should be placed in `/test/migrations/` folder. Most probably
```
```suggestion
Every mgration must have test coverage. Tests for migrations should be placed in the `/test/migrations/` folder. Most probably
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+* `/src/repo/lock-memory.js`:`lock()` - locking repository that uses memory
+
+Every mgration has to have a test coverage. Tests for migration should be placed in `/test/migrations/` folder. Most probably
+you will have to plug the tests into `browser.js`/`node.js` if they require specific bootstrap on each platform.
```
```suggestion
you will have to plug the tests into `browser.js`/`node.js` if they require specific bootstrapping on each platform.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+#### Empty migrations
+
+For inter-operable reasons with Go-IPFS it might be necessary just to bump a version of repo without any actual 
```
```suggestion
For interop with go-ipfs it might be necessary just to bump a version of a repo without any actual 
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+#### Empty migrations
+
+For inter-operable reasons with Go-IPFS it might be necessary just to bump a version of repo without any actual 
+modification as there might not be any changes needed in JS implementation. For that you can create "empty migration".
```
```suggestion
modification as there might not be any changes needed in the JS implementation. For that purpose you can create an "empty migration".
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+For inter-operable reasons with Go-IPFS it might be necessary just to bump a version of repo without any actual 
+modification as there might not be any changes needed in JS implementation. For that you can create "empty migration".
+
+The easiest way is to use the CLI for that:
```
```suggestion
The easiest way to do so is with the CLI:
```

dirkmc commented on README.md:
````diff
@@ -1 +1,284 @@
+> jsipfs-migrations add --empty
+```
+
+This will create empty migration with next version in line.
````
```suggestion
This will create an empty migration with the next version.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+### `migrate(path, {toVersion, ignoreLock, repoOptions, onProgress, isDryRun}) -> Promise<void>`
+
+Executes forward migration to specific version or if not specified to the latest version.
```
```suggestion
Executes a forward migration to a specific version, or to the latest version if a specific version is not specified.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+ * `path` (string, mandatory) - path to the repo to be migrated
+ * `options` (object, optional) - options for the migration
+ * `options.toVersion` (int, optional) - version to which the repo should be migrated to. If left out the version of latest migration is used.
```
```suggestion
 * `options.toVersion` (int, optional) - version to which the repo should be migrated. Defaults to the latest migration version.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+ * `path` (string, mandatory) - path to the repo to be migrated
+ * `options` (object, optional) - options for the migration
+ * `options.toVersion` (int, optional) - version to which the repo should be migrated to. If left out the version of latest migration is used.
+ * `options.ignoreLock` (bool, optional) - if true won't lock the repo for applying the migrations. Use with caution.
```
```suggestion
 * `options.ignoreLock` (bool, optional) - if true will not lock the repo when applying migrations. Use with caution.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+ * `options` (object, optional) - options for the migration
+ * `options.toVersion` (int, optional) - version to which the repo should be migrated to. If left out the version of latest migration is used.
+ * `options.ignoreLock` (bool, optional) - if true won't lock the repo for applying the migrations. Use with caution.
+ * `options.repoOptions` (object, optional) - options that are passed to migrations, that use them to correctly construct datastore. Options are same like for IPFSRepo.
```
```suggestion
 * `options.repoOptions` (object, optional) - options that are passed to migrations, that use them to construct the datastore. (options are the same as for IPFSRepo).
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+ * `options.ignoreLock` (bool, optional) - if true won't lock the repo for applying the migrations. Use with caution.
+ * `options.repoOptions` (object, optional) - options that are passed to migrations, that use them to correctly construct datastore. Options are same like for IPFSRepo.
+ * `options.onProgress` (function, optional) - callback that is called after finishing execution of each migration to report progress.
+ * `options.isDryRun` (bool, optional) - flag that indicates if it is a dry run that should imitate running migration without actually any change.
```
```suggestion
 * `options.isDryRun` (bool, optional) - flag that indicates if it is a dry run that should give the same output as running a migration but without making any actual changes.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+**Arguments:**
+ * `migration` (object) - object of migration that just successfully finished running. See [Architecture of migrations](#architecture-of-migrations) for details.
+ * `counter` (int) - current number of migration in the planned migrations streak.
+ * `totalMigrations` (int) - total count of migrations that are planned to be run.
```
```suggestion
 * `totalMigrations` (int) - total count of migrations that will be run.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+**Arguments:**
+ * `migration` (object) - object of migration that just successfully finished running. See [Architecture of migrations](#architecture-of-migrations) for details.
+ * `counter` (int) - current number of migration in the planned migrations streak.
```
```suggestion
 * `counter` (int) - index of current migration.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+### `revert(path, toVersion, {ignoreLock, options, onProgress, isDryRun}) -> Promise<void>`
+
+Executes backward migration to specific version.
```
```suggestion
Executes backward migration to a specific version.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+ * `path` (string, mandatory) - path to the repo to be reverted
+ * `toVersion` (int, mandatory) - version to which the repo should be reverted to. 
+ * `options` (object, optional) - options for the reversion
+ * `options.ignoreLock` (bool, optional) - if true won't lock the repo for applying the migrations. Use with caution.
```
```suggestion
 * `options.ignoreLock` (bool, optional) - if true will not lock the repo when applying migrations. Use with caution.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+ * `toVersion` (int, mandatory) - version to which the repo should be reverted to. 
+ * `options` (object, optional) - options for the reversion
+ * `options.ignoreLock` (bool, optional) - if true won't lock the repo for applying the migrations. Use with caution.
+ * `options.options` (object, optional) - options that are passed to migrations, that use them to correctly construct datastore. Options are same like for IPFSRepo.
```
```suggestion
 * `options.options` (object, optional) - options that are passed to migrations, that use them to construct the datastore. (options are the same as for IPFSRepo).
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+ * `options.ignoreLock` (bool, optional) - if true won't lock the repo for applying the migrations. Use with caution.
+ * `options.options` (object, optional) - options that are passed to migrations, that use them to correctly construct datastore. Options are same like for IPFSRepo.
+ * `options.onProgress` (function, optional) - callback that is called after finishing execution of each migration to report progress.
+ * `options.isDryRun` (bool, optional) - flag that indicates if it is a dry run that should imitate running migration without actually any change.
```
```suggestion
 * `options.isDryRun` (bool, optional) - flag that indicates if it is a dry run that should give the same output as running a migration but without making any actual changes.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+ 
+### `getLatestMigrationVersion() -> int`
+
+Return the version of latest migration.
```
```suggestion
Return the version of the latest migration.
```

dirkmc commented on README.md:
```diff
@@ -1 +1,284 @@
+
+## CLI
+
+The package comes also with CLI that is exposed as NodeJS binary with name `jsipfs-repo-migrations`. 
```
```suggestion
The CLI is a NodeJS binary named `jsipfs-repo-migrations`. 
```

dirkmc commented:
I'll continue reviewing the code tomorrow - looking good!

AuHau commented:
Thanks for the grammar check 😅 It is highly appreciated!

AuHau commented:
I have also moved the repo under IPFS org. What kind of teams should be assigned to it? I have added JS Team, but not sure if some others should have access also.

Also what about "Lead Maintainer"? I can do it, or if somebody else should be it (because PL guidelines or something), then I don't mind that either.

dirkmc commented on run.md:
```diff
@@ -0,0 +1,41 @@
+
+This document explains how to run [repo](https://github.com/ipfs/specs/tree/master/repo/) migrations for [js-ipfs](https://github.com/ipfs/js-ipfs).
+
+Note that running migrations is a task automatically performed by the `js-ipfs` when running the `js-ipfs` command after installing its new version, so you would normally not need to run the `jsipfs-repo-migrations` tool.
```
```suggestion
Note that the `js-ipfs` command automatically runs migrations when a new version is installed, so you would normally not need to run the `jsipfs-repo-migrations` tool.
```

dirkmc commented on run.md:
```diff
@@ -0,0 +1,41 @@
+
+Note that running migrations is a task automatically performed by the `js-ipfs` when running the `js-ipfs` command after installing its new version, so you would normally not need to run the `jsipfs-repo-migrations` tool.
+
+The `jsipfs-repo-migrations` comes into play when the internal, on-disk format `js-ipfs` uses to store data changes. In order to avoid losing data, this tool upgrades old versions of the repo to the new ones.
```
```suggestion
`jsipfs-repo-migrations` comes into play when the internal, on-disk format `js-ipfs` uses to store data changes. To enable the new version of `js-ipfs` to read the new data format, this tool upgrades from an old version of the repo to a new version.
```

dirkmc commented on run.md:
```diff
@@ -0,0 +1,41 @@
+
+## Step 1. Downloading the Migration
+
+Recommended way is to use `npm` and install this package with it: `npm install ipfs-repo-migrations`
```
```suggestion
`npm install ipfs-repo-migrations`
```

dirkmc commented on run.md:
```diff
@@ -0,0 +1,41 @@
+
+## Step 2. Run the Migration
+
+Now, run the migration tool:
```
```suggestion
To run the migration tool:
```

dirkmc commented on run.md:
````diff
@@ -0,0 +1,41 @@
+Now, run the migration tool:
+
+```sh
+# you can check if there are migrations that needs to be applied
````
```suggestion
# to check if there are migrations that need to be applied
```

dirkmc commented on run.md:
```diff
@@ -0,0 +1,41 @@
+# you can check if there are migrations that needs to be applied
+js-ipfs-repo-migrations status
+
+# if so, you can migrate to the latest version
```
```suggestion
# if so, migrate to the latest version
```

dirkmc commented on src/cli.js:
```diff
@@ -0,0 +1,74 @@
+
+const commands = require('./commands')
+
+function print (msg, newline) {
```
```suggestion
function print (msg = '', newline = true) {
```

dirkmc commented on src/cli.js:
```diff
@@ -0,0 +1,74 @@
+        throw err // preserve stack
+      }
+
+      if (args.length > 0) {
```
Why only print the message `if (args.length > 0)`?

AuHau replied:
This is to suppress message `You need at least one command before continuing` when you run the command without any parameters and which display the help page.

I took most of this file from `js-ipfs` I think.

dirkmc replied:
Makes sense 👍 
Please add a comment to the code explaining that

dirkmc commented on src/cli.js:
```diff
@@ -0,0 +1,74 @@
+      print('Unknown error, please re-run the command with DEBUG=repo-migrations:cli to see debug output')
+    }
+
+    exitCode = 1
```
You can probably just remove the `exitCode` altogether, and move process.exit(1) inside this catch block

AuHau replied:
Makes sense 👍

dirkmc commented on src/commands.js:
```diff
@@ -0,0 +1,157 @@
+  } else if (revertOk) {
+    await migrator.revert(repoPath, to, options)
+  } else {
+    throw new Error('The migration would require reversion of the repo, but you have not allowed it as \'--revert-ok\' is not present.')
```
```suggestion
    throw new Error(`The migration would revert the repo version from ${version} to ${to}. To revert you must supply '--revert-ok'.`)
```

dirkmc commented on src/commands.js:
```diff
@@ -0,0 +1,157 @@
+  const statusString =
+    version < lastMigrationVersion ? chalk.yellow('There are migrations to be applied!') : chalk.green('Nothing to migrate!')
+
+  return `${statusString}\nCurrent repo version: ${version}\nLast migration's version: ${lastMigrationVersion}`
```
```suggestion
  return `${statusString}\nCurrent repo version: ${version}\nLatest migration version: ${lastMigrationVersion}`
```

dirkmc commented on src/commands.js:
```diff
@@ -0,0 +1,157 @@
+  const version = await repoVersion.getVersion(repoPath)
+  const lastMigrationVersion = migrator.getLatestMigrationVersion(migrations)
+  const statusString =
+    version < lastMigrationVersion ? chalk.yellow('There are migrations to be applied!') : chalk.green('Nothing to migrate!')
```
```suggestion
    version < lastMigrationVersion ? chalk.yellow('Repo is out of date') : chalk.green('Nothing to migrate')
```

dirkmc commented on src/commands.js:
```diff
@@ -0,0 +1,157 @@
+  const newMigrationVersion = migrator.getLatestMigrationVersion() + 1
+  const newMigrationFolder = path.join(__dirname, '..', 'migrations', 'migration-' + newMigrationVersion)
+
+  const migrationsImport = migrations.map((migration) => migration.empty ? `  Object.assign({version: ${migration.version}}, emptyMigration),` : `  require('./migration-${migration.version}'),`)
```
Let's make sure migration.version is a number here

AuHau replied:
I don't think that is needed. This generates string so even if `migration.version` would be a string, it would not make any difference.

dirkmc replied:
I'm thinking more along the lines of code injection issues

dirkmc commented on src/commands.js:
```diff
@@ -0,0 +1,157 @@
+    handler: asyncClosure(migrate),
+    builder: yargv => yargv
+      .option('to', {
+        describe: 'Specify to which version should be repo migrated to',
```
```suggestion
        describe: 'Target version to migrate repo to',
```

dirkmc commented on src/commands.js:
```diff
@@ -0,0 +1,157 @@
+        type: 'number'
+      })
+      .option('dry', {
+        describe: 'Only displays what migrations will be reverted',
```
```suggestion
        describe: `Output migration results but don't actually perform migration`,
```

dirkmc commented on src/commands.js:
```diff
@@ -0,0 +1,157 @@
+        type: 'boolean'
+      })
+      .option('revert-ok', {
+        describe: 'Allows to do backward migration, if the specific version is lower the current version of repository',
```
```suggestion
        describe: 'Allow migrating to a lower version (reverting)',
```

dirkmc commented on src/commands.js:
```diff
@@ -0,0 +1,157 @@
+  },
+  status: {
+    command: 'status',
+    describe: 'Display status of IPFS repo',
```
```suggestion
    describe: 'Display migration status of IPFS repo',
```

dirkmc commented on src/cli.js:
```diff
@@ -0,0 +1,68 @@
+const commands = require('./commands')
+
+function print (msg = '', newline = true) {
+  if (newline === undefined) {
```
My suggestion to change the line above from
`print (msg, newline)` => `print (msg = '', newline = true)`
was so that you could remove lines 13 - 19

AuHau replied:
😆 yeah I did not realize that, I just blindly accepted your suggestion. Will fix.

dirkmc commented on src/index.js:
```diff
@@ -0,0 +1,235 @@
+      try {
+        if (!isDryRun) await migration.migrate(path, repoOptions, isBrowser)
+      } catch (e) {
+        e.message = `During migration to version ${migration.version} exception was raised: ${e.message}`
```
Can we use [explain-error](https://www.npmjs.com/package/explain-error) here so as not to lose the stack trace?

AuHau replied:
Well, of course, we can, but I am not sure how much of added value will it bring. The stack traces are persisted using this approach.
I looked into how `explain-error` generates stack traces and at least to me, they seems super confusing. Bellow is how stack trace looks like with this approach:
```
  repo-migrations:repo_fs_lock locking /Users/adam/.jsipfs/repo.lock +0ms
  repo-migrations:migrator Migrating version 8 +0ms
  repo-migrations:repo_fs_lock releasing lock /Users/adam/.jsipfs/repo.lock +4ms
  repo-migrations:cli {
  repo-migrations:cli   error: Error: During migration to version 8 exception was raised: Some err
  repo-migrations:cli       at Object.migrate (/Users/adam/Projects/ipfs/js-ipfs-repo-migrations/migrations/migration-8/index.js:55:9)
  repo-migrations:cli       at Object.migrate (/Users/adam/Projects/ipfs/js-ipfs-repo-migrations/src/index.js:99:40)
  repo-migrations:cli       at async migrate (/Users/adam/Projects/ipfs/js-ipfs-repo-migrations/src/commands.js:67:5)
  repo-migrations:cli       at async main (/Users/adam/Projects/ipfs/js-ipfs-repo-migrations/src/cli.js:52:22),
  repo-migrations:cli   argv: {
  repo-migrations:cli     _: [ 'migrate' ],
  repo-migrations:cli     resolve: [Function],
  repo-migrations:cli     reject: [Function],
  repo-migrations:cli     '$0': 'src/cli.js'
  repo-migrations:cli   }
  repo-migrations:cli } +0ms
During migration to version 8 exception was raised: Some err
```

dirkmc replied:
Ok if you think the error will be clear enough let's leave it 👍

dirkmc commented on src/index.js:
```diff
@@ -0,0 +1,235 @@
+    }
+
+    if (!isDryRun) await repoVersion.setVersion(path, toVersion || getLatestMigrationVersion(migrations))
+    log('All migrations successfully migrated ', toVersion !== undefined ? `to version ${toVersion}!` : 'to latest version!')
```
```suggestion
    log('Repo successfully migrated ', toVersion !== undefined ? `to version ${toVersion}!` : 'to latest version!')
```

dirkmc commented on src/index.js:
```diff
@@ -0,0 +1,235 @@
+
+  let reversibility = verifyReversibility(migrations, currentVersion, toVersion)
+  if (!reversibility.reversible) {
+    throw new errors.NonReversibleMigrationError(`Migration version ${reversibility.version} is not possible to revert! Cancelling reversion.`)
```
```suggestion
    throw new errors.NonReversibleMigrationError(`It is not possible to revert to version ${toVersion} because migration version ${reversibility.version} is not reversible. Cancelling reversion.`)
```

dirkmc commented on src/index.js:
```diff
@@ -0,0 +1,235 @@
+      try {
+        if (!isDryRun) await migration.revert(path, repoOptions, isBrowser)
+      } catch (e) {
+        e.message = `During reversion to version ${migration.version} exception was raised: ${e.message}`
```
Can we use explain-error here too?

dirkmc commented on src/index.js:
```diff
@@ -0,0 +1,235 @@
+  }
+
+  if (migrationCounter !== (fromVersion - toVersion)) {
+    throw new errors.NonReversibleMigrationError(`There are missing migration to perform the reversion! 
```
```suggestion
    throw new errors.NonReversibleMigrationError(`There are not enough migrations to perform the reversion! 
```

dirkmc commented on src/index.js:
```diff
@@ -0,0 +1,235 @@
+
+  if (migrationCounter !== (fromVersion - toVersion)) {
+    throw new errors.NonReversibleMigrationError(`There are missing migration to perform the reversion! 
+    Expected ${(fromVersion - toVersion)} migrations, but there is only ${migrationCounter} migrations.`)
```
```suggestion
    Expected ${(fromVersion - toVersion)} migrations, but there are only ${migrationCounter} migrations.`)
```

dirkmc commented on src/repo/init.js:
```diff
@@ -0,0 +1,31 @@
+
+    return true
+  } catch (e) {
+    log('While checking if repo is initialized error was thrown: ' + e.message)
```
Please use explain-error here

dirkmc commented on src/repo/version.js:
```diff
@@ -0,0 +1,48 @@
+    throw new errors.UnknownRepoStructureError('Repo does not have version file! Is the repo initialized?')
+  }
+
+  const version = parseInt(await store.get(versionKey))
```
It seems like [fs-datastore returns a Buffer](https://github.com/ipfs/js-datastore-fs/blob/master/src/index.js#L188) from `store.get()`

Should the result first be converted to a string before calling `parseInt()`?

AuHau replied:
Well, I believe that `parseInt()` does this implicitly, so I would say no need for it. If for code clarity you think it should be explicit, the let me know.

dirkmc replied:
Interesting, ok I didn't think that through :)

AuHau replied:
Tbh, I did not think about it before you mentioned, but [documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt#Parameters) confirmed it ;-)

dirkmc commented:
@AuHau thank you again for all your hard work. This is looking really good 👍 

@alanshaw I've done a couple of passes over the code and suggested some changes, all of which @AuHau has addressed. I think this is ready for you to review 🚀

daviddias commented on .editorconfig:
```diff
@@ -0,0 +1,6 @@
+indent_size = 2
+trim_trailing_whitespace = true
+insert_final_newline = true
+end_of_line = lf
```
Mind adding this file to .gitignore?

AuHau replied:
Well I can, but I think it is a good thing to have around as it unities the code style. I know that for that there is `aegir lint` but this integrates directly with IDEs. You don't use it in other projects? I feel like I took it from some other JS package...
For example here: https://github.com/ipfs/js-ipfs/blob/master/.editorconfig

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+ * Handles locking/unlocking of repository
+ * Defines migrations API
+ * Executes and reports migrations in both directions: forward and backward
+ * Simplifies creation of new migrations
```
```suggestion
 * Simplifies creation of new migrations
 * Works on the browser too!
```

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+
+##### `migrate(repoPath, isBrowser)`
+
+_Do not confuse this function with the `require('ipfs-repo-migrations').migrate()` function that drives the whole migration process!_
```
After reading this, I am now confused. What is the different then?

AuHau replied:
Damn 😂 

There are two functions with the same name in this tool:

 - root one (eq. `require('ipfs-repo-migrations').migrate()`) which is part of the main API, that upon calling will migrate the repository to new version
 - part of "migration API" that is present on the objects exported in each migration for example in `require('ipfs-repo-migrations/migrations/migration-8').migrate()` that only migrate forward the repo by one version

Do you have some idea how to make it more clear?

daviddias replied:
It makes more sense if you move the API to the top, so that is the first thing the user sees. Otherwise it is information overload and the user didn't even get the time to test the module.

AuHau replied:
Yup, I have done that. It is definitely better flow like this...

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+migrations leaves the user with no way to run the migrations because there is no CLI in the browser. In such
+a case, you should provide a way to trigger migrations manually.**
+
+### Writing migration
```
```suggestion
## Creating a new migration
```

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+are reverted before propagating the error. If possible then the outcome should be consistent repo so it migration could
+be run again. 
+
+#### Architecture of migrations
```
```suggestion
### Architecture of a migration
```

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+ * `migrate` (function) - Function that performs the migration (see signature of this function below)
+ * `revert` (function) - If defined then this function will revert the migration to the previous version. Otherwise it is assumed that it is not possible to revert this migration.
+
+##### `migrate(repoPath, isBrowser)`
```
```suggestion
#### `.migrate(repoPath, isBrowser)`
```

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+ * `options` (object, optional) - object containing `IPFSRepo` options, that should be used to construct a datastore instance.
+ * `isBrowser` (bool) - indicates if the migration is run in a browser environment (as opposed to NodeJS)
+ 
+##### `revert(repoPath, isBrowser)`
```
```suggestion
#### `.revert(repoPath, isBrowser)`
```

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+ * `options` (object, optional) - object containing `IPFSRepo` options, that should be used to construct the datastore instance.
+ * `isBrowser` (bool) - indicates if the migration is run in a browser environment (as opposed to NodeJS)
+
+#### Browser vs. NodeJS environments
```
This could be done automatically, no? All it should care about is the path

AuHau replied:
Yes, as mentioned lower in most of the cases the difference is handled by datastore implementation. But there might be a need for this in some more complicated migrations.

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+ So with simple migrations you shouldn't worry about the difference between `datastore-fs` and `datastore-level` 
+ and by default use the `datastore-fs` package (as the replace mechanism does not work vice versa).
+
+#### Guidelines
```
```suggestion
### Guidelines
```

daviddias commented on README.md:
````diff
@@ -1 +1,293 @@
+The recommended way to write a new migration is to first bootstrap a dummy migration using the CLI:
+
+```sh
+> jsipfs-migrations add
````
This requires for folks to install the module globally. I would rather suggest it as a npm script

```suggestion
> npm run new-migration
```

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+A new folder is created with the bootstrapped migration. You can then simply fill in the required fields and 
+write the rest of the migration! 
+
+#### Migration's dependencies
```
This section can be removed. I don't think it introduces any new concept to avid js developers that will be doing something as advanced as a repo migration.

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+An update of some dependency could introduce breaking change. In such a case the next steps should be discussed with a broader 
+audience. 
+
+#### Integration with js-ipfs
```
```suggestion
### Integration with js-ipfs
```

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+When a new migration is created, the repo version in [`js-ipfs-repo`](https://github.com/ipfs/js-ipfs-repo) should be updated with the new version,
+together with updated version of this package. Then the updated version should be propagated to `js-ipfs`.
+
+#### Tests
```
```suggestion
### Tests
```

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+Every migration must have test coverage. Tests for migrations should be placed in the `/test/migrations/` folder. Most probably
+you will have to plug the tests into `browser.js`/`node.js` if they require specific bootstrapping on each platform.
+
+#### Empty migrations
```
```suggestion
### Empty migrations
```

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+|                  7 | v0.0.0 - latest  |
+
+
+## API
```
The API of the module should come before the explanation of how to create a new migration as that is how most people will use the module

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+ * `options.toVersion` (int, optional) - version to which the repo should be migrated. Defaults to the latest migration version.
+ * `options.ignoreLock` (bool, optional) - if true will not lock the repo when applying migrations. Use with caution.
+ * `options.repoOptions` (object, optional) - options that are passed to migrations, that use them to construct the datastore. (options are the same as for IPFSRepo).
+ * `options.onProgress` (function, optional) - callback that is called after finishing execution of each migration to report progress.
```
Is this for progress bars?

AuHau replied:
Yup

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+
+## API
+
+### `migrate(path, {toVersion, ignoreLock, repoOptions, onProgress, isDryRun}) -> Promise<void>`
```
```suggestion
### `.migrate(path, {toVersion, ignoreLock, repoOptions, onProgress, isDryRun}) -> Promise<void>`
```

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+ * `counter` (int) - index of current migration.
+ * `totalMigrations` (int) - total count of migrations that will be run.
+
+### `revert(path, toVersion, {ignoreLock, options, onProgress, isDryRun}) -> Promise<void>`
```
```suggestion
### `.revert(path, toVersion, {ignoreLock, options, onProgress, isDryRun}) -> Promise<void>`
```

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+ 
+For further details see the `--help` pages.
+
+## Versioning
```
```suggestion
## Developer

### Module versioning notes 
```

daviddias commented on README.md:
````diff
@@ -1 +1,293 @@
+
+```js
+const migrations = require('ipfs-repo-migrations')
+const getVersion = require('ipfs-repo-migrations/repo/version')
````
This is an actual antipattern. Teaching users to require specific files will make them fall on their face as a simple file name change can go unnoticed and break everyone's usage. Instead, use `migrations.getVersion`

AuHau replied:
Fair point. Yet the `getVersion` function is there only for CLI sake and is not part of the main API. I wrote it here only to make the example more clear. I would suggest then only rewrite the example?

alanshaw replied:
For this example I'd just use `const currentRepoVersion = 7`

daviddias commented on README.md:
```diff
@@ -1 +1,293 @@
+
+To migrate your repository using the CLI, see the [how to run migrations](./run.md) tutorial. 
+
+**For tools that build on top of `js-ipfs` and run mainly in the browser environment, be aware that disabling automatic
```
This should go into js-ipfs docs :)

alanshaw commented on README.md:
```diff
@@ -1 +1,277 @@
+ * `counter` (int) - index of current migration.
+ * `totalMigrations` (int) - total count of migrations that will be run.
+
+### `.revert(path, toVersion, {ignoreLock, options, onProgress, isDryRun}) -> Promise<void>`
```
Rename `options` -> `repoOptions`?

alanshaw commented on README.md:
```diff
@@ -1 +1,277 @@
+
+Each migration must follow this API. It must export an object in its `index.js` that has following properties:
+
+ * `version` (int) - Number that represents the version which the repo will migrate to (eg. `migration-8` will move the repo to version 8).
```
`migration-8` -> `8` - it is a number not a string?

AuHau replied:
Good catch. The property is a number, but the migrations are in folder `migration-<migration version number>`. I will rephrase this.

alanshaw commented on README.md:
```diff
@@ -1 +1,277 @@
+ * `migrate` (function) - Function that performs the migration (see signature of this function below)
+ * `revert` (function) - If defined then this function will revert the migration to the previous version. Otherwise it is assumed that it is not possible to revert this migration.
+
+#### `.migrate(repoPath, isBrowser)`
```
Missing `options` from signature

alanshaw replied:
Can we detect `isBrowser` using https://github.com/ipfs/js-ipfs-utils/blob/71a7bf54ab3636bae01da80d5a73aa7f839668ff/src/env.js#L6 instead?

alanshaw commented on README.md:
```diff
@@ -1 +1,277 @@
+Arguments:
+ * `repoPath` (string) - absolute path to the root of the repo
+ * `options` (object, optional) - object containing `IPFSRepo` options, that should be used to construct a datastore instance.
+ * `isBrowser` (bool) - indicates if the migration is run in a browser environment (as opposed to NodeJS)
```
Should `isBrowser` be an option?

AuHau replied:
You mean as part of `options`/future `repoOptions`? If so, then I don't think so, because these options should be linked only to constructions of datastore and should follow the same semantic as the object passed to `IPFSRepo`...

alanshaw commented on README.md:
```diff
@@ -1 +1,277 @@
+### Integration with js-ipfs
+
+When a new migration is created, the repo version in [`js-ipfs-repo`](https://github.com/ipfs/js-ipfs-repo) should be updated with the new version,
+together with updated version of this package. Then the updated version should be propagated to `js-ipfs`.
```
I think some clarification is needed here. I'm assuming `ipfs-repo` depends on `ipfs-repo-migrations`? So the order should be to publish a new version of `ipfs-repo-migrations` and then `ipfs-repo`?

Would we then never actually use the "update to latest" functionality, because the repo would have to specify what version to up/down-grade to? Either that or we need to be really careful with the npm version range for `ipfs-repo-migrations` so that it doesn't pull in a version that can upgrade a repo to a version higher than what is supported.

For example.

1. I have `ipfs-repo@0.28.0` which operates on repo v7 and I depend on `ipfs-repo-migrations@^1.0.0`
2. Then I want to upgrade the repo to v8 so do the work in `ipfs-repo` and write a migration
3. I release `ipfs-repo@0.29.0` and `ipfs-repo-migrations@^1.1.0`

So `ipfs-repo` needs to specify the version to migrate to, or `ipfs-repo@0.28.0` will try to migrate to v8 when it only actually supports v7 because of the `^` version range on the `ipfs-repo-migrations` dep.

In summary: what I'm suggesting is that the call to `migrator.migrate` **requires** a version to up/down-grade to. 😁

AuHau replied:
Good catch with the npm version range! Currently `ipfs-repo` always sets to which version should the repo be migrated to, so it makes sense to make it as mandatory. For the "migrate to latest" feature, which is mainly needed for CLI we can easily call the `getLatestMigrationVersion()`.

alanshaw commented on README.md:
```diff
@@ -1 +1,277 @@
+
+### Module versioning notes 
+
+In order to have good overview of what version of package contains what kind of migrations, this package follows this versioning schema: `0.<versionOfLastMigration>.<patches>`.
```
Tying the npm version to the repo version like this means we lose the ability to communicate breaking/feature change information for _this_ module to dependants.

For example, how would we communicate a change like the async/await refactor in this module? There's no repo version increase and we can't release a breaking change like that in a patch release.

AuHau replied:
Well, you are right about this, but I don't believe there will be much of feature development/breaking changes in the future of this package. I believe that 80% of the changes will be adding new migrations, then 15% will be fixing bugs and 5% others...

In your example, we could communicate the breaking change using the major version. So for example from `0.8.0` we could bump to `1.8.0`... 
I kind of like the idea that you know what migrations the package contains just from the version number.

alanshaw replied:
> In your example, we could communicate the breaking change using the major version. So for example from `0.8.0` we could bump to `1.8.0`...

We could...IMHO that would be a really strange versioning strategy though!

> I kind of like the idea that you know what migrations the package contains just from the version number.

I kind of do too, but I'm not convinced it's worth it!

AuHau replied:
> We could...IMHO that would be a really strange versioning strategy though!

True, but following `0.<versionOfLastMigration>.<patches>` is already "strange" anyway ;-) 

What about having two parts in the version? E.g. `<semver>-<last version of migration>` so it could look like `0.2.0-7`? When new migration is added it would be bumped to `0.3.0-8`. This would allow us to have full-fledged semver capabilities and still have the information about migrations as part of the version?

What I am not sure is if this schema would not break some tooling? (aegir?)

alanshaw replied:
Is there another reason other than convenience that would warrant this versioning strategy? 

If you don't already know the module the versioning strategy isn't going to convey any additional info. Secondly, it's really easy for new developers to miss this important piece of info since it's a small section at the bottom of the README. Thirdly, if you do know the module and are upgrading to get the latest migrations you'll probably just need the latest version of this module anyway right? It's quite unlikely you'll ever need an older version of the module since the newer versions contain all the past migrations as well.

AuHau replied:
I would not call it convenience, but being careful. Having a different version of this package installed may lead to irreversible actions that can happen without your knowledge (eq. when you have `autoMigrations` on - which is currently the default). Yes, it won't most probably lose the data per se, but it might migrate data that are not to be migrated at that moment (for example you have test repo and "production" repo and you want to do some tests, but you accidentally migrate wrong one etc...).

Of course, having this schema won't prevent all mix ups, but it can improve the situation. From version update `0.4.0 --> 0.5.0` you can't be sure if there was new migration added (yes, it will be most of the cases), or maybe some feature change? You will have to go study release notes, which should be always the case, but I am pretty sure a lot of devs don't do that. 

In this light, I think having the `-<latestVersionOfMigration>` suffix is even better than the original schema as it stands out from Semver and should tingle dev's senses about what it most probably means.

I am not sure if my thoughts about this are not getting too tangled and strayed from reality, but I still feel like there are more pros over cons with this schema. Actually are there any cons? Except for the possible tooling side?

alanshaw replied:
If you feel strongly about this I'll not stand in your way any further.

I'm undecided if I agree that a pre-release suffix is a better strategy than using the regular major/minor/patch versions. My issue with this it is that according to semver:

> A pre-release version indicates that the version is unstable and might not satisfy the intended compatibility requirements as denoted by its associated normal version

So this would convey to devs that it is not a stable version that should not be used without accepting some instability consequences.

I also do not know if AEgir will support you on this 😛

AuHau replied:
Hmmm. Agree that this is a bit unfortunate. I checked finally SemVer spec and it directly supports ["build metadata"](https://github.com/ipfs/js-ipfs-repo-migrations/pull/4), which I feel could be exactly used for this. It stands aside in determining version precedence, which is a good thing.

We could have then schema `<semver>+<latestMigrationVersion>` (i.e `0.1.0+7`) to meet both points... Only thing I am not sure is about the format of the metadata. Should it be just integer of the version or somethings like `+migration.<int>`? It would be more self-explanatory, but longer...

alanshaw commented on run.md:
````diff
@@ -0,0 +1,41 @@
+The migration tool is safe -- it should not delete any data. If you have important data stored _only_ in your ipfs node, and want to be extra safe, you can back up the whole repo with:
+
+```sh
+cp -r ~/.js-ipfs ~/.js-ipfs.bak
````
Suggested change:

```sh
cp -r ~/.jsipfs ~/.jsipfs.bak
```

alanshaw commented on run.md:
```diff
@@ -0,0 +1,41 @@
+js-ipfs-repo-migrations status
+
+# if so, migrate to the latest version
+js-ipfs-repo-migrations migrate
```
In the README doc I think you reference the CLI tool command as `jsipfs-repo-migrations`

alanshaw commented on package.json:
```diff
@@ -0,0 +1,75 @@
+    "url": "https://github.com/ipfs/js-ipfs-repo-migrations/issues/"
+  },
+  "license": "MIT",
+  "leadMaintainer": "Adam Uhlir <adam@uhlir.dev>",
```
FYI we use `maintainer.json` for this info now e.g. https://github.com/ipfs/js-ipfs/blob/master/maintainer.json

alanshaw commented on src/index.js:
```diff
@@ -0,0 +1,245 @@
+  }
+
+  if (!(await repoInit.isRepoInitialized(path))) {
+    throw new errors.NotInitializedRepoError(`Repo in path ${path} is not initialized!`)
```
Push the message into the error constructor?

```js
throw new errors.NotInitializedRepoError(path)
```

alanshaw commented on src/index.js:
```diff
@@ -0,0 +1,245 @@
+const repoLock = require('./repo/lock')
+const repoInit = require('./repo/init')
+const isBrowser = require('./option-node')
+const errors = require('./errors')
```
Suggestion: destructure the errors you're using?

AuHau replied:
Well, to be honest, I learned in Python not to do this. First of all, it gives more context about the function/error (not really applicable in this case) and when you add new function/error into the module you don't have to think about adding it in the import. Also when you remove something you don't have to again think about removing it from import...

alanshaw commented on src/index.js:
```diff
@@ -0,0 +1,245 @@
+  const currentVersion = await repoVersion.getVersion(path)
+
+  if (currentVersion === toVersion) {
+    log('Nothing to migrate, skipping migrations.')
```
Nit, not really skipping anything here!

alanshaw commented on src/index.js:
```diff
@@ -0,0 +1,245 @@
+  }
+
+  if (currentVersion > toVersion) {
+    log(`Current repo's version (${currentVersion}) is higher then toVersion (${toVersion}), nothing to migrate.`)
```
Should be an error IMHO since it will not leave the repo in the state the caller expects.

alanshaw commented on src/index.js:
```diff
@@ -0,0 +1,245 @@
+        throw e
+      }
+
+      typeof onProgress === 'function' && onProgress(migration, counter, totalMigrations) // Reports on migration process
```
Nit, personally I'd default this to a noop function (`onProgress = onProgress || (() => {})`) instead of checking every time and using &&. It makes this line a little more readable and less confusing to more junior programmers.

alanshaw commented on src/index.js:
```diff
@@ -0,0 +1,245 @@
+
+  let reversibility = verifyReversibility(migrations, currentVersion, toVersion)
+  if (!reversibility.reversible) {
+    throw new errors.NonReversibleMigrationError(`It is not possible to revert to version ${toVersion} because migration version ${reversibility.version} is not reversible. Cancelling reversion.`)
```
It would be cool if the CLI got you to confirm before performing a migration that was not reversable.

alanshaw commented on src/repo/lock.js:
```diff
@@ -0,0 +1,38 @@
+ * would throw an exception because it couldn't update the lock file mtime within
+ * the 10s threshold. @see https://github.com/ipfs/js-ipfs-repo/pull/182
+ * Increasing to 20s is a temporary fix a permanent fix should be implemented in
+ * the future.
```
I believe this problem has been resolved so no need to alter this stale timeout.

AuHau replied:
Well, this was directly taken from https://github.com/ipfs/js-ipfs-repo/blob/master/src/lock.js so I am a bit confused. It is still there...

Should I remove this custom STALE_TIME?

alanshaw replied:
I think that is also out of date - it was fixed in proper-lockfile here I believe https://github.com/moxystudio/node-proper-lockfile/pull/74
